### PR TITLE
SI-7470 implements fundep materialization

### PR DESCRIPTION
**Pull request by xeno-by**
review @adriaanm @odersky @retronym @paulp

### CODE REVIEW

**Comment by xeno-by** [comment]
This pull request builds upon #2494, enabling implicit macros to affect type inference. Unfortunately we had a disagreement about whether we want to include this functionality in trunk, so I moved the controversial bits into a separate pull request. The difference between #2494 and this pull request is a single commit: https://github.com/scalamacros/kepler/compare/ticket/5923...topic/iso.

**Comment by xeno-by** [comment]
**The motivation**

```
trait Iso[T, U] {
  def to(t : T) : U
  def from(u : U) : T
}
implicit def materializeIso[T, U]: Iso[T, U] = macro ???
```

We have a multi-parametric `Iso[T, U]` type class, which represents isomorphisms between case classes (`T`) and HLists (`U`). Okay, so we have a fundep `T -> U`, and now we want to materialize that fundep. 

**The problem** 

```
case class Foo(i: Int, s: String, b: Boolean)
def foo[C, L](c: C)(implicit iso: Iso[C, L]): L = iso.to(c)
foo(Foo(23, "foo", true))
```

We cannot use the regular materialization technique, because there's a problem with type inference. As shown on the snippet above, when inferring an implicit argument for an invocation of foo, we look for an implicit of type `Iso[C, ?L]`. However macros cannot expand when a macro application contains undetparams, so before expanding the implicit macro, the macro engine will instantiate undetparams, turning `?L` into `Nothing`. And then even if `materializeIso` expands into `new Iso[C, Int :: String :: Boolean :: HNil] { ... }`, it wouldn't typecheck against `Iso[C, Nothing]`, which makes materialization of fundep multiparametric type classes impossible.

**The solution**

The patch consists of two simple ideas, both having just a handful of locs in their implementation:
1) Even if a macro application contains undetparams, let it expand. 
2) When typechecking the result of macro expansion, replace all undetparams in the expected type with wildcards.

In the `Iso[C, ?L]` case, the materialization macro doesn't care about `?L` being not inferred, because it knows exactly what to infer it to. So it will simply expand into `new Iso[C, Int :: String :: Boolean :: HNil] { ... }` and be done with the it. When typechecking the expansion, macro engine typechecks it against `Iso[C, ?]` instead of `Iso[C, ?L]`, which works out. And now type inference sees that the type of the implicit candidate is `Iso[C, Int :: String :: Boolean :: HNil]`, so it figures out the correct value for `?L`.

**tl;dr**

Even if you can't infer all the type parameters of implicit being looked up, let the macro expand. All the undetparams will later be inferred from the result of expansion. The implementation of that technique is extremely simple - just 8 lines of code.

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala [hunk 1]
```diff
@@ -706,6 +706,11 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
 
             var expectedTpe = expandee.tpe
             if (isNullaryInvocation(expandee)) expectedTpe = expectedTpe.finalResultType
+            var undetparams = List[Symbol]()
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:709]: This should probably use `approximateAbstracts`. `undetparams` is not used except for replacing them with wildcards, so it should be fine.

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala [hunk 2]
```diff
@@ -706,6 +706,11 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
...
+            def traverse(sym: Symbol) = if (sym.isTypeParameter && !sym.isSkolem) undetparams ::= sym
+            expectedTpe.foreach(sub => traverse(sub.typeSymbol))
+            expectedTpe = deriveTypeWithWildcards(undetparams)(expectedTpe)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:712]: Can this be expressed with `expectedTpe` a `val`, not a `var`?

Let's call the approximated expected type `ptApprox`

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala [hunk 3]
```diff
@@ -757,8 +762,21 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
...
+          // and then trigger macro expansion with the undetermined type parameters still there.
+          // Thanks to that the materializer can take a look at what's going on and react accordingly.
+          // Simple, yet powerful and much more lightweight and elegant than onInfer.
+          // (What's onInfer? See http://docs.scala-lang.org/overviews/macros/inference.html)
           val shouldInstantiate = typer.context.undetparams.nonEmpty && !inPolyMode(mode)
-          if (shouldInstantiate) typer.instantiatePossiblyExpectingUnit(delayed, mode, pt)
+          if (shouldInstantiate) {
+            forced += delayed
+            typer.infer.inferExprInstance(delayed, typer.context.extractUndetparams(), pt, keepNothings = false)
+            macroExpand(typer, delayed, mode, WildcardType)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:778]: Shouldn't this use at least `ptApprox` as expected type? (Or `pt` with `undetparams` replaced by their inferred types.)

**Comment by adriaanm** [comment]
- [ ] Please document the test case a bit with what's being tested.

**Comment by adriaanm** [comment]
- [ ] one more test case

Also, more convincing examples would help support this PR's case. A common use case would be in database mappers where the companion object is boilerplate.

```
case class RowA extends MappedRow
implicit object RowA extends RowMapper[RowA]

case class RowB extends MappedRow
implicit object RowB extends RowMapper[RowB]
```

could be shortened to 

```
case class RowA extends MappedRow
case class RowB extends MappedRow
```

if the library provides a macro like

```
implicit macro def mapper[T <: MappedRow]: RowMapper[T] = macro mapperImpl
def mapperImpl = reify { new RowMapper[T] }
```

**Comment by adriaanm** [comment]
Finally, we should probably reserve new features for 2.11.
2.10.x is the stable release, it's not supposed to get new features anymore.

**Comment by xeno-by** [comment]
Thank you for the feedback! I'll resubmit once I address it.

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala (hunk 1 above)
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:709]: See https://github.com/scala/scala/pull/2821/files#L0L599.

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:709]: I also tried to change `approximateAbstracts` to work for this case and see what breaks. You can follow the experiment live at https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/987/console.
